### ISSUE TITLE
ActiveStorage accepts broken images and throws exceptions when creating variants

### ISSUE THREAD

**Issue by Signum** [rails/rails#40493]
**TL;DR** There is no way to detect broken image files. So exceptions are thrown later when ActiveStorage tries to render _variants_.

### Steps to reproduce

Create a controller action that receives uploaded image files. A part from my Rails application:

```ruby
def upload_receive
  params[:file].each do |img|
    new_screenshot = @package.screenshots.new(image: img)
    
    if new_screenshot.valid?  
      new_screenshot.save
    end
  end
end
```

My model:

```ruby
class Screenshot < ApplicationRecord
  has_one_attached :image
  validates :image, attached: true, content_type: [:png]

  def medium_image
    if self.image.attached?
      self.image.variant(
        resize_to_limit: [670, 600]
      ).processed
    end
  end
end
```

(The validation uses the 'active_storage_validations' gem but the problem is most certainly not coming from there.)

This works for intact PNG files. A deliberately broken PNG file will be accepted and saved though. My test file is 5 KB of zero bytes that I named "broken.png" and uploaded to Rails.

### Expected behavior

One of…

1. ActiveStorage detect broken image files by trying to open them with Minimagick/Imagemagick as a validation.
1. ActiveStorage offers a common validation to detect broken images. (I tried to write my own custom validator to run the file through Minimagick but it I could not find a way to access the blob's data in the model before a ".save".)
1. Creating variants does not just throw an exception.

### Actual behavior

The file gets saved by ActiveStorage with content_type='image/png'. I suspect that 'marcel' and 'mimemagick' (which I believe are used for the content type detection) can't figure out what this file is and just look at the '.png' suffix.

Once a template tries to render a variant using the Screenshot.medium_image method it will call Minimagick and fail:

```
convert /tmp/ActiveStorage-21670-… failed with error:
convert-im6.q16: improper image header `/tmp/ActiveStorage-21670-…' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: no images defined `/tmp/image_processing….png' @ error/convert.c/ConvertImageCommand/3258.
```

### System configuration

**Rails version**: 6.0.3.4

**Ruby version**: 2.6.5p114

**Comment by tahsin352** [comment]
I guess you are using this https://github.com/igorkasyanchuk/active_storage_validations validation. It handles this validation for uploaded images.
```ruby
validates :image, attached: true, content_type: [:png]
```
Regarding this broken image validation please create an issue on https://github.com/igorkasyanchuk/active_storage_validations/issues

**Comment by fatkodima** [comment]
> The file gets saved by ActiveStorage with content_type='image/png'. I suspect that 'marcel' and 'mimemagick' (which I believe are used for the content type detection) can't figure out what this file is and just look at the '.png' suffix.

This is how `marcel` actually detects the mime-type: 
https://github.com/basecamp/marcel/blob/cce58f26c728980a2e1f9ba5fe0b11aedf43b283/lib/marcel/mime_type.rb#L15-L24

It looks for the "magic" header first, then name, then extension, etc. The `blobs.content_type` column's value is identified and set by ActiveStorage. The `active_storage_validations` then just basically string-compares the expected value with the column's value.

I can not answer why `marcel` detects the mime-type the way it does, but I would implement it by detecting by magic header and then fallback to binary. 

For your use case, I think one of the solutions would be to monkey-patch `marcel` with my suggestion, something like
```ruby
    def for(pathname_or_io = nil, name: nil, extension: nil, declared_type: nil)
      type_from_data = for_data(pathname_or_io)
      fallback_type = for_declared_type(declared_type) || for_name(name) || for_extension(extension) || BINARY

      if type_from_data
        most_specific_type type_from_data, fallback_type
      else
        BINARY
      end
    end
```

**Comment by brendon** [comment]
This is still an issue. In my case a user uploaded a 0 byte image file. ActiveStorage happily accepted it and happily went to create a variant, then just blew up with:

```
ActionView::Template::Error admin/newsletter_sets/online_newsletters/articles/images#create_many
VipsForeignLoad: "/tmp/ActiveStorage-614-20230531-50206-aexfrd.jpg" is not a known file format
```

The problem becomes annoying because it basically kills access to the page where the offending image could be removed to fix the problem meaning I need to go in and delete it from the console.

In my case I'll add an extra check on the file size to make sure it's more than `0` but ideally guards should be in place at the ActiveStorage level to prevent a file from getting classified as an image if it's not. I know there are some guards but obviously not one for this.

**Comment by Mth0158** [comment]
Hi,
I am a maintainer of the [active_storage_validation](https://github.com/igorkasyanchuk/active_storage_validations) gem and I have been stumbling on this problem lately. The `VipsForeignLoad` can be raised when reading a 0 byte size file, as mentioned at https://github.com/janko/image_processing/issues/97.

This edge case works fine in with `MiniMagick`, but it does not work properly with `Vips`, using active_storage_validation gem you can handle this use case with the `processable_image` validator. I'll fix this Vips issue in the active_storage_validation gem in the coming days and it should be released soon.

**Comment by cseelus** [comment]
Also a problem for invalid images dragged from Microsoft Teams (`MicrosoftTeams-image-n.png`). Haven't found a solution to filter them out yet.

Could be handled a bit more gracefully from Rails imo, will look into that if I find the time.
